ndk/src/hardware_buffer.rs

```diff
@@ -73,12 +79,317 @@ impl HardwareBufferUsage {
+        }
+    }
+
+    #[cfg(feature = "api-level-29")]
+    pub fn lock_planes(
+        &self,
+        usage: HardwareBufferUsage,
+        fence: Option<RawFd>,
+        rect: Option<Rect>,
+    ) -> Result<HardwareBufferPlanes> {
+        let fence = fence.unwrap_or(-1);
+        let rect = match rect {
+            Some(rect) => &rect,
+            None => std::ptr::null(),
+        };
+        let planes = construct(|res| unsafe {
+            ffi::AHardwareBuffer_lockPlanes(self.as_ptr(), usage.0, fence, rect, res)
+        })?;
+
+        Ok(HardwareBufferPlanes {
+            inner: planes,
+            index: 0,
+        })
+    }
+
+    pub fn unlock(&self) -> Result<()> {
+        let status = unsafe { ffi::AHardwareBuffer_unlock(self.as_ptr(), std::ptr::null_mut()) };
+        if status == 0 {
+            Ok(())
+        } else {
+            Err(HardwareBufferError(status))
+        }
+    }
+
+    /// Returns a fence file descriptor that will become signalled when unlocking is completed,
+    /// or `None` if unlocking is already finished.
+    pub fn unlock_async(&self) -> Result<Option<RawFd>> {
+        let fence = construct(|res| unsafe { ffi::AHardwareBuffer_unlock(self.as_ptr(), res) })?;
+        Ok(match fence {
+            -1 => None,
+            fence => Some(fence),
+        })
+    }
+
+    pub fn recv_handle_from_unix_socket(socket_fd: RawFd) -> Result<Self> {
+        unsafe {
+            let ptr =
+                construct(|res| ffi::AHardwareBuffer_recvHandleFromUnixSocket(socket_fd, res))?;
+
+            Ok(Self::from_ptr(NonNull::new_unchecked(ptr)))
+        }
+    }
+
+    pub fn send_handle_to_unix_socket(&self, socket_fd: RawFd) -> Result<()> {
+        unsafe {
+            let status = ffi::AHardwareBuffer_sendHandleToUnixSocket(self.as_ptr(), socket_fd);
+            if status == 0 {
+                Ok(())
+            } else {
+                Err(HardwareBufferError(status))
+            }
+        }
+    }
+
+    pub fn acquire(&self) -> HardwareBufferRef {
+        unsafe {
+            ffi::AHardwareBuffer_acquire(self.as_ptr());
+        }
+        HardwareBufferRef {
+            inner: HardwareBuffer { inner: self.inner },
+        }
+    }
+}
+
+/// A `HardwareBuffer` with an owned reference, the reference is released when dropped.
+/// It behaves much like a strong `Rc` reference.
+#[derive(Debug)]
+pub struct HardwareBufferRef {
+    inner: HardwareBuffer,
+}
+
+impl Deref for HardwareBufferRef {
+    type Target = HardwareBuffer;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl Drop for HardwareBufferRef {
+    fn drop(&mut self) {
+        unsafe {
+            ffi::AHardwareBuffer_release(self.inner.as_ptr());
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct HardwareBufferDesc {
+    width: u32,
+    height: u32,
+    layers: u32,
+    format: HardwareBufferFormat,
+    usage: HardwareBufferUsage,
+    stride: u32,
```

Comment:
@JasperDeSutter While writing/importing documentation for `hardware_buffer` and referring to one of these fields I found "by accident" that they are private. Is that intended?

Reply:
Heh, yeah that would make a lot more sense! I guess I didn't need it for my use case where the information is passed through jni, and simply forgot to add the pubs (no beer, I swear 😄)

Reply:
No worries, thanks for confirming!